ic-cdk/src/call.rs

```diff
@@ -432,6 +432,91 @@ pub struct CandidDecodeFailed {
     candid_error: String,
 }
 
+/// Extension trait for error types to provide additional methods.
+pub trait CallErrorExt {
+    /// Checks if the error is a clean reject.
+    /// A clean reject means that there must be no state changes on the callee side.
+    fn is_clean_reject(&self) -> bool;
+    /// Checks if the failed call is immediately retryable.
+    /// This method will return false if the call is likely to
+    /// just fail again if we retry it immediatly.
+    fn is_immediately_retryable(&self) -> bool;
+}
+
+impl CallErrorExt for CallPerformFailed {
+    fn is_clean_reject(&self) -> bool {
+        true
+    }
+
+    fn is_immediately_retryable(&self) -> bool {
+        // TODO: check if this is correct
+        false
+    }
+}
+
+impl CallErrorExt for CallRejected {
+    fn is_clean_reject(&self) -> bool {
+        // Here we apply a conservative whitelist of reject codes that are considered clean.
+        // Once finer `error_code` is available, we can allow more cases to be clean.
+        matches!(
+            self.reject_code,
+            RejectCode::SysFatal | RejectCode::SysTransient | RejectCode::DestinationInvalid
+        )
+    }
+
+    fn is_immediately_retryable(&self) -> bool {
+        // Here we apply a conservative whitelist of reject codes that are considered clean.
```

Comment:
Can't leave a suggestion for some reason:
        // Here we apply a whitelist of reject codes that are considered immediately retryable.